src/python/pants/util/meta.py

```diff
@@ -111,7 +112,47 @@ def staticproperty(func: Union[staticmethod, Callable]) -> ClassPropertyDescript
   return ClassPropertyDescriptor(func, doc)
 
 
-def frozen_after_init(cls: Type[T]) -> Type[T]:
+class _ClassDecoratorWithSentinelAttribute(ABC):
+  """Base class to wrap a class decorator which sets a "sentinel attribute".
+
+  This functionality is exposed via the `@sentinel_attribute` decorator.
+  """
+
+  @abstractmethod
+  def __call__(self, cls: type) -> type: ...
+
+  def define_instance_of(self, obj: type, **kwargs) -> type:
+    return type(obj.__name__, (obj,), {
+      '_sentinel_attribute_type': type(self),
+      **kwargs
+    })
```

Comment:
MyPy is cool with this?

Reply:
Yes!! It's just returning a type -- is there something suspicious about it?